Fix scheduler issues caused by compound triggers

eric-wieser opened the pull request:
This is an alternative to #642

This adds support for a new type of object to the scheduler, a `Waitable`.
A `Waitable` is something that produces a new coroutine every time it is yielded.
`yield some_waitable` is nothing more than syntactic sugar for `yield some_waitable._wait()`.

With this new primitive, we can:
* Fix ClockCycles (gh-520)
* Fix Combine (gh-852)
* Remove `yield [a, b]` as a primitive, and build it into a new `yield First(a, b)` waitable. This allows a reasonable amount of simplification to the scheduler.

imphil commented:
@TC01 @putsplut could you have a look at this code and see if it solves the problems that motivated #642 (and consequently in #538)? I didn't take a closer look at the code yet, but I generally like the approach and the simplifications it brings.

eric-wieser commented:
Rebased and fixed a comment.

I considered including `Sequential` here, written as:
```python
class Sequential(_AggregateWaitable):
    def _wait(self):
        for trig in self.triggers:
            yield trig
        raise ReturnValue(self)
```

However, I'm not really sure it's a useful addition. It can always be added in a later PR.

TC01 commented:
Thanks for submitting this!

I'll take a look and test it. I think that if my testbenches work properly using this patch then I have no objections to using it in favor of the hierarchical triggers branch-- I agree it's nice to have a simpler solution. :)

TC01 commented:
So, I no longer get the cbValueChange message from trying to schedule ClockCycles and RisingEdge at the same time, which is good! At least some of my tests now run with master + this branch.

Unfortunately, one of them seems to hang. That test doesn't do anything too unusual compared to the ones that aren't hanging, either, so... I suspect there's a race condition somewhere.

It's possible there was just a change of behavior that caused some assumption I made somewhere to break? I'll have to debug this more, see if I can find out exactly what ```yield``` statement is causing a hang (my testbench setup makes heavy use of forked coroutines, and communication between them, so it will be a bit difficult to trace through) and report back.

eric-wieser commented:
It's likely some scheduler fixes I made recently have had an effect here.

I don't expect the rebase to be easy either for that reason.

eric-wieser commented:
Turns out this also fixes #gh-843

imphil commented on cocotb/triggers.py:
```diff
@@ -544,3 +463,130 @@ def prime(self, callback):
+    Combines multiple triggers together.
+
+    Waits until the first trigger fires, and returns its result.
+    """
```
Please also document what happens to triggers that didn't fire. And what is the behavior if two triggers fire at the same time?

eric-wieser replied:
I don't think "at the same time" is a well-defined concept in an event loop - at best there's a concept of "within the same simulation tick", but the events still fire in a defined order.

imphil replied:
So what is the defined order from a user perspective? And please add a note that the non-first waitables are being killed.

cmarqu replied:
I have another small stylistic comment - in https://www.python.org/dev/peps/pep-0257/ they say 
> [The docstring is a phrase ending in a period.] It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

eric-wieser replied:
> And please add a note that the non-first waitables are being killed.



They're not - killing isn't transitive, the only thing that is killed is the internal coroutine used to wait on the things passed in. Killing a coroutine does not kill any nested calls currently in progress.

eric-wieser replied:
@cmarque: I'd argue that this is not a function docstring, so it doesn't make sense to document as a command.

eric-wieser commented on cocotb/scheduler.py:
```diff
@@ -490,13 +489,17 @@ def _event_loop(self, trigger):
         """Unschedule a coroutine.  Unprime any pending triggers"""
 
-        for trigger in self._coro2triggers[coro]:
+        # unprime the trigger if any are left
```
Now just one, will fix

eric-wieser commented:
`Combine` is a bad name, but it's now a new one - that trigger already existed.
Perhaps we should deprecate it and rename to `All` in a later PR

imphil commented:
Ah yes, I forgot that we had `Combine` before. Let's deprecate/rename it in a follow-up.

eric-wieser commented:
Documentation comments addressed.

TC01 commented:
> It's likely some scheduler fixes I made recently have had an effect here.

I think I tracked this down, after I found a second test that was failing. It looks like there was a logic path in which I was doing this:

```
while True:
    yield event.wait()
    if some_condition is True:
         do_something()
        event.clear()
```

This will *not* always clear the event trigger before attempting to yield it again-- which was probably an oversight on my part. But it looks like the scheduler's behavior has changed; previously it seems like this worked. Modifying the code to always call ```event.clear()``` before looping seems to resolve the problem though.

I am just running my full set of tests again to make sure there are no other issues. :)

eric-wieser commented:
I think some thought might be needed about whether First() does the right thing for such a useful name.

For instance, what should `First(First(a, b), c)` return if `b` fires?

imphil commented:
> For instance, what should First(First(a, b), c) return if b fires?

`b`? (Given that b fires first)

eric-wieser commented:
Or should it return `First(b, c)`, since that fires first?

imphil commented:
CI fails for 

```
Traceback (most recent call last):
  File "/src/cocotb/__init__.py", line 41, in <module>
    from cocotb.scheduler import Scheduler
  File "/src/cocotb/scheduler.py", line 67, in <module>
    from cocotb.triggers import (Trigger, GPITrigger, Timer, ReadOnly, PythonTrigger,
  File "/src/cocotb/triggers.py", line 576
    else:
       ^
IndentationError: expected an indented block
```

imphil commented:
> Or should it return First(a, b), since that fires first?

What would a use case for that be?

But maybe there is, what I had in mind (return `b`) can already be written as `First(a, b, c)`.

(Another option would be to disallow this syntax for now until we have a clear use case for it.)

TC01 commented:
Just a quick update: my testbench seems to be working fine with this branch now. I think I'm fine with dropping the hierarchical triggers branch and using this instead.

imphil commented:
Thanks for your feedback Ben!

eric-wieser commented:
> What would a use case for that be?

Generic code written with the assumption that `(yield First(*triggers)) in triggers` is always true. Perhaps there's another way to handle that though.

eric-wieser commented:
Updated to fix CI, and with a test for `First(First(a, b), c)`. Great to hear this solved your issues, @TC01!

eric-wieser commented on cocotb/scheduler.py:
```diff
@@ -454,9 +454,8 @@ def _event_loop(self, trigger):
+                    self._coro2trigger[coro]
                     for coro in scheduling
-                    for t in self._coro2triggers[coro]
                 } - {trigger}
```
Looking more closely, this entire thing is dead code now. Do you want me to do another rev to remove it, or leave it for a future PR?

eric-wieser replied:
Whatever, fix was easy, and this removed even more lines

eric-wieser commented on cocotb/scheduler.py:
```diff
@@ -450,22 +450,6 @@ def _event_loop(self, trigger):
-                    t
-                    for coro in scheduling
-                    for t in self._coro2triggers[coro]
-                } - {trigger}
```
This set was now always empty

eric-wieser commented:
> So what is the defined order from a user perspective

It's up to the simulator. The point is that the simulator tells us about events one at a time, there is no such thing as simultaneous events.

imphil commented:
> > So what is the defined order from a user perspective
> 
> It's up to the simulator. The point is that the simulator tells us about events one at a time, there is no such thing as simultaneous events.

Even though the simulator and the scheduler code doesn't use the term "simultaneous events", there are still such events from a user's perspective: Think about two `Timer(1000)` triggers -- which one fires first? I'd love to see documentation helping a user to understand what happens in this scenario: that it is simulator-dependent behavior.

eric-wieser commented:
Does that documention belong in First, or elsewhere?

eric-wieser commented:
Added a `.. note::` in `First`. I don't want to spend much more time documenting this, especially since `First` is not a new feature, just a new way to spell an old one.

imphil commented:
Thanks a lot @eric-wieser for going the extra mile on this PR regarding documentation. Your scheduler work is highly appreciated and the deep understanding you've brought into this area of the code has made cocotb significantly better. Thanks!
